workflow.md

```diff
@@ -15,95 +17,226 @@
 * Features, fixes, refactoring must be done in separate branches.
-  * They are reviewed and merged after that.
-* New tagged release (even minor one) is a new branch.
-* If there is a release then a `tag` is created from the commit of this release
-  branch.
-* We use release branches for GitHub Releases page, please see [one](https://github.com/input-output-hk/cardano-sl/releases) for Cardano SL project.
-* Fixes from release branches are merged into `master`.
+* If there is a release then a `tag` is created from the commit
+  corresponding to this release.
+* We use release branches for GitHub Releases page, please see
+  [one](https://github.com/input-output-hk/cardano-sl/releases) for
+  `cardano-sl` project.
+* Release branches must be merged into `master` every time changes
+  happen there (single commit or a small group of related commits
+  which happened in a short period of time). It's important, because
+  otherwise `master` eventually differs from release branch too much
+  and nobody can merge it. You can pass `-s ours` to `git merge` to
+  discard all changes if it's desirable. Such merges must be done via
+  pull requests too. `master` is never merged into release branch.
+* `cardano-sl` project uses a more complicated workflow with `-mid`
+  and `-rc` branches. `-mid` branches are intended to be used by other
+  projects which depend on `cardano-sl`. `-rc` branch is created
+  prior to release. These branches are protected too. It's described
+  in more detail [below](#special-workflow-for-cardano-sl-project).
+* If a project depends on a concrete version (specified as a commit
+  hash) of another project, this version must be in **protected**
+  branch. It ensures that commit with given hash will not disappear.
 
 ## Branches
 
-* Branch name should fit the following template: `(type of task)/(maybe related
-  issue number)(name)`.
-  * `bugfix/dae31-fix-tx-send`
-  * `feature/csl492-wallet-delegation-certs`
-  * `feature/launcher`
+* Generally branch name should fit the following template:
+  `(nickname)/(maybe related issue number-)(very brief
+  description)`. If nickname is not applicable to a branch (most likely
+  because multiple people works there) it can be skipped. Examples:
+  * `akegalj/dae31-fix-tx-send`
+  * `csl492-wallet-delegation-certs`
+  * `gromak/launcher`
 
 ## Commits
 
-We follow [this guide](http://chris.beams.io/posts/git-commit/) how to write a
-commit message. Also if commit is related to an issue then we add the issue name
-at the beggining (i.e. `[CSL-659] Introduce Pos.Communication.Protocol`).
+We follow [this guide](http://chris.beams.io/posts/git-commit/) how to
+write a commit message. Also if commit is related to an issue then one
+must add the issue identifier at the beggining (e. g. `[CSL-659] Introduce
+Pos.Communication.Protocol`). Issue identifier is not necessary only
+in few cases, specifically:
+* in merge commits;
+* if commit doesn't modify AST of the program;
```

Comment:
Abstract syntax tree. For example, when you add a comment or split one line into multiple lines, etc.

I don't know how to formulate it better. I think that we also don't need a special issue if we just fix a typo in a variable name which is used only in 1 or 2 places, for example. But I don't know how to describe it formally.